### ISSUE TITLE
javax.annotation.processing.FilerException: Attempt to reopen a file for path /META-INF/additional-spring-configuration-metadata.json

### ISSUE THREAD

**Issue by kerwin612** [egoettelmann/spring-configuration-extensions#10]
```bash
[WARNING] Unable to merge additional metadata
[WARNING] javax.annotation.processing.FilerException: Attempt to reopen a file for path /.cache/act/f8cc9a9e9eafffbb/hostexecutor/target/classes/META-INF/additional-spring-configuration-metadata.json
  	at jdk.compiler/com.sun.tools.javac.processing.JavacFiler.checkFileReopening(JavacFiler.java:766)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacFiler.getResource(JavacFiler.java:598)
  	at org.springframework.boot.configurationprocessor.MetadataStore.getAdditionalMetadataStream(MetadataStore.java:104)
  	at org.springframework.boot.configurationprocessor.MetadataStore.readAdditionalMetadata(MetadataStore.java:76)
  	at org.springframework.boot.configurationprocessor.ConfigurationMetadataAnnotationProcessor.mergeAdditionalMetadata(ConfigurationMetadataAnnotationProcessor.java:350)
  	at org.springframework.boot.configurationprocessor.ConfigurationMetadataAnnotationProcessor.writeMetadata(ConfigurationMetadataAnnotationProcessor.java:339)
  	at org.springframework.boot.configurationprocessor.ConfigurationMetadataAnnotationProcessor.process(ConfigurationMetadataAnnotationProcessor.java:189)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1023)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.runContributingProcs(JavacProcessingEnvironment.java:859)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1265)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1404)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1234)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:916)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
  	at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:136)
  	at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:183)
  	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1140)
  	at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:193)
  	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:126)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:342)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:330)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:213)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:175)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.access$000(MojoExecutor.java:76)
  	at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:163)
  	at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute(DefaultMojosExecutionStrategy.java:39)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:160)
  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:105)
  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:73)
  	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:53)
  	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:118)
  	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:261)
  	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:173)
  	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:101)
  	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:827)
  	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:272)
  	at org.apache.maven.cli.MavenCli.main(MavenCli.java:195)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
```

**Comment by egoettelmann** [comment]
Thanks for the issue and the related PR.
I am trying to reproduce, but unfortunately I have not enough inputs regarding the context to do so.
Also, I do not see how the PR is related or resolves the encountered issue.

Do you provide a [custom Spring metadata file](https://docs.spring.io/spring-boot/docs/current/reference/html/configuration-metadata.html#appendix.configuration-metadata.annotation-processor.adding-additional-metadata) ?

Do you encounter this stacktrace during your regular Maven build or through your IDE ?

What is the build environment: Java version, OS ?

Many thanks in advance for your response

**Comment by kerwin612** [comment]
Apache Maven 3.8.8 (4c87b05d9aedce574290d1acc98575ed5eb6cd39)
Maven home: apache-maven-3.8.8
Java version: 17.0.7, vendor: Oracle Corporation, runtime: jdk-17.0.7
Default locale: zh_CN, platform encoding: UTF-8
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

**I located that the exception occurred because `maven-compiler-plugin` used version `3.11.0`;**  

You just need to add the following code under [this line of code](https://github.com/egoettelmann/spring-configuration-extensions/blob/fb8809d2fc4eca832075147adf44d413abb5f12d/spring-configuration-extensions-samples/pom.xml#L77) to reproduce the problem.
```xml
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.11.0</version>
            </plugin>
```

-----  

I debugged this exception is that the **spring-value-annotation-processor** opened the **META-INF/additional-spring-configuration-metadata.json**  
An exception was thrown when **spring-boot-configuration-processor** attempted to read the **META-INF/additional-spring-configuration-metadata.json**: **reopen**;     

I tracked that all **processor** shares a common context, so this problem seems unsolvable!(Because two **processor**'s read and write the same file **META-INF/additional-spring-configuration-metadata.json** together)     

When the user does not use the **META-INF/additional-spring-configuration-metadata.json**, this exception will not have any impact, but if the user uses it, the problem will be exposed, and the **spring-value-annotation-processor** will write the parsed @value to **META-INF/additional-spring-configuration-metadata.json**, resulting in the loss of the user defined **META-INF/additional-spring-configuration-metadata.json** content!

-----  

My solution is to not read and write the same file(**META-INF/additional-spring-configuration-metadata.json**) with **spring-value-annotation-processor** together!   
**And I believe that even if this exception does not occur, allowing users to customize the path of related files is also a good feature!**

**Comment by kerwin612** [comment]
You can refer to [the example of this branch](https://github.com/kerwin612/spring-configuration-extensions/blob/patch-2/spring-configuration-extensions-samples), only the **META-INF\additional-spring-configuration-metadata.json** file is added to the original example;   
The difference between **3.11.0** and **3.1** of **maven-compiler-plugin** versions is that **3.11.0** will output an error log; **3.1** does not output error logs;  
The contents of the final custom file are overwritten by **spring-value-annotation-processor**, instead of merging together as expected.

**Comment by egoettelmann** [comment]
Many thanks for the detailed description of the problem and of course for the fix.
All changes have been merged into `develop` and released in version `0.1.6`.

However, I changed the proposed namings into:
- compiler arg of `spring-value-annotation-processor` becomes (`metadataOutputFile`):
  ```
  <compilerArgs>
    <arg>-AmetadataOutputFile=${your-custom-file-path}</arg>
  </compilerArgs>
  ```
- configuration option of `spring-configuration-aggregator-maven-plugin` becomes:
  ```
  <additionalMetadataFiles>
    <metadataFile>${your-custom-file-path}</metadataFile>
  </additionalMetadataFiles>
  ```

The README files have been updated accordingly, and a *Known issues* section has been added to reference this issue.

Also, I reworked the samples module to integrate more easily different use cases with different sub-modules. The `sample-2` provides an example with the usage of the custom path.
